aws/cl_zcash/software/runtime/ecdsa_test.cpp

```diff
@@ -0,0 +1,333 @@
+//
+//  ZCash FPGA test.
+//
+//  Copyright (C) 2019  Benjamin Devlin and Zcash Foundation
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+#define _XOPEN_SOURCE 500
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdarg.h>
+#include <assert.h>
+#include <string.h>
+#include <string>
+
+#include <unistd.h>
+#include <stdlib.h>
+
+#include <fpga_pci.h>
+#include <fpga_mgmt.h>
+#include <utils/lcd.h>
+#include <utils/sh_dpi_tasks.h>
+
+#include "zcash_fpga.hpp"
+
+#include <openssl/ecdsa.h>
+#include <openssl/sha.h>
+#include <openssl/bn.h>
+#define NID NID_secp256k1   //using NID_secp256k1:714
+
+typedef struct __attribute__((__packed__)) {
+    uint64_t s[4];
+    uint64_t r[4];
+    uint64_t hash[4];
+    uint64_t Qx[4];
+    uint64_t Qy[4];
+} Signature_t;
+
+/* use the stdout logger for printing debug information  */
+
+const struct logger *logger = &logger_stdout;
+/*
+ * check if the corresponding AFI for hello_world is loaded
+ */
+//int check_afi_ready(int slot_id); //I think this is not necessary as it is already override..
+
+
+void usage(char* program_name) {
+    printf("usage: %s [--slot <slot-id>][<poke-value>]\n", program_name);
+}
+
+// uint32_t byte_swap(uint32_t value);
+
+uint32_t byte_swap(uint32_t value) {
+    uint32_t swapped_value = 0;
+    int b;
+    for (b = 0; b < 4; b++) {
+        swapped_value |= ((value >> (b * 8)) & 0xff) << (8 * (3-b));
+    }
+    return swapped_value;
+}
+
+bool string_to_hex(const std::string &inStr, unsigned char *outStr) {
+    size_t len = inStr.length();
+    for (ssize_t i = len-2; i >= 0; i -= 2) {
+        sscanf(inStr.c_str() + i, "%2hhx", outStr);
+        ++outStr;
+    }
+    return true;
+}
+
+
+Signature_t sig_ossl (bool verb)	{	//verb=true, print more info.
```

Comment:
I think this function (and maybe some other code) is duplicated between files, can you put them into a shared lib or common file?

Reply:
Will do another version update.